### PR TITLE
MOAPT-2671 - Update Jira plugin for the new Atlassian board view.

### PR DESCRIPTION
**Pull request by Alex-McEvoy**
### Jira URL
[cds-global.atlassian.net/browse/MOAPT-2671](https://cds-global.atlassian.net/browse/MOAPT-2671)

### Description
Atlassian is forcing all users of Jira to use the new look on Nov 5th.

Update the Jira plugin that shows PR labels for the new Atlassian view, which changes the page structure significantly.
The new look removes card elements from the DOM as they are scrolled out of view, and re-renders them when they are back in view. This is a shift from the previous way that cards were rendered, and requires us to listen for new cards creation.

1. Update css selectors for pertinent elements. 
2. Use MutationObservers to listen for card creation when scrolling
3. Cache previously observed card info, to speed up label insertion.

### QA Instructions
I'm not sure if there's a better way of doing this. Let me know if you know of one.
Get a zip file from Alex, and unzip it somewhere on your computer. You'll use that in step 10.

1. Make sure you're in the new look in Jira 
Screen Shot 2023-10-23 at 11 53 48 AM

2. unzip the downloaded zip file sent to you

3. Go to the chrome extensions manager 

Screen Shot 2023-10-23 at 11 41 29 AM

4. Turn off the existing extension `Github Labels for JIRA Sprints 0.14.0`

Screen Shot 2023-10-23 at 11 41 56 AM

5. click on the "Load Unpacked" button 

Screen Shot 2023-10-23 at 11 43 09 AM

6. Select the unzipped folder that you unzipped earlier, the whole folder, with everything in it should be called something like `gh-labels-in-jira`

Screen Shot 2023-10-23 at 11 44 47 AM

7. Be sure the extension is turned on, and the version is `0.14.1`

Screen Shot 2023-10-23 at 11 46 09 AM

8. Go to the sprint board (or any sprint board) https://cds-global.atlassian.net/jira/software/c/projects/MOAPT/boards/275, and see that the extension is there, you might need to click on the extension icon, and pin the plugin

Screen Shot 2023-10-23 at 11 47 27 AM

9. click on the green icon in the toolbar, and select "options"

Screen Shot 2023-10-23 at 11 49 07 AM

10. you'll need a github access token for this part, there's instructions on how to generate a new one here. https://www.notion.so/mediaos/Github-Labels-in-Jira-b05f734721b8463f9b79adab7a2ba462?pvs=4#d41996adfbeb471e828548e74e7268db
After you're done, put that token in the field in the plugin here. It's recommended to keep this token in a secure place, because you might need it again if you install/uninstall the plugin more than once, and you won't be able to get it from github after you copy it initially. 

Screen Shot 2023-10-23 at 11 47 59 AM

11. Go back to the sprint board, and refresh the page, you should see the labels being applied

Screen Shot 2023-10-23 at 11 52 27 AM

12. Be sure to check the teams kanban board as well. https://cds-global.atlassian.net/jira/software/c/projects/MOAPT/boards/276 Make sure you've added a column name from the kanban board in the "options" of the ticket, an example might be "review needed". (it's not case sensitive anymore). You can also select "Show PRs in all columns" to see all PRs associated with all tickets.

13. Be sure to go back to the "options" by clicking on the plugin, and test the functionality of all the various buttons there. `Hide labels on closed PR's`, `Hide closed PR's` etc.

### Acceptance Criteria
Given that I navigate to a Jira view for a given board, when I look a the tickets that have a pull request associated, then I would see the Github labels for that ticket displayed below each ticket.

### CODE REVIEW

File: content.js [hunk 1]
```diff
@@ -300,39 +275,106 @@ function addCodeReviewers(owner, repo, prid, label_id, requested_reviewers, labe
...
+    const newCards = [...mutations].filter(
+        mutationRecord => mutationRecord?.addedNodes?.[0]?.childNodes?.[1]?.id?.includes("card-")
+        ).map(mutation => mutation?.addedNodes?.[0]?.childNodes?.[1]);
+    
+    newCards.forEach(cardNode => {
+        if (awesome_loading) {
+            loadAwesomely(cardNode);
+        }
+        
+        if (MEMOIZED_CARDS[cardNode]) {
```
Comment (sondoselayyanOP) [inline content.js]:
```suggestion
        if (MEMOIZED_CARDS[cardNode?.id]) {
```

**Review by sondoselayyanOP — CHANGES_REQUESTED** [review]

File: content.js (hunk 1 above)
Comment (sondoselayyanOP) [reply content.js]: Question: should use the id of cardNode instead of cardNode because we are setting the cardId as the key in MEMOIZED_CARDS?

Comment (Alex-McEvoy) [reply content.js]: ahh yes, good catch! Updated

**Comment by sondoselayyanOP** [comment]
@Alex-McEvoy I tried to active the `Hide labels on closed PR's` and `Hide closed PR's` , but they didn't work

**Comment by jana-998** [comment]
@Alex-McEvoy  I am having the same problem as Sondos, for `Hide labels on closed PR's` and `Hide closed PR's` 
Other wise LGTM!

**Comment by Alex-McEvoy** [comment]
@sondoselayyanOP @jana-998 Hey thanks for taking a look all! Good catch on the hiding PRs feature.

So it looked like before, the `Hide closed PR's` and the `Hide labels on closed PR's` buttons did essentially the same thing, just hid "DECLINED" PRs. Hiding labels is the same thing as hiding PRs. Note that closed PRs here didn't mean "MERGED" PRs.

I've added another button to hide "MERGED" PRs, and removed the redundant `Hide labels on closed PR's`. The updates should hide both "DECLINED" and "MERGED" PRs, according to the user selection. 

It might be helpful to either add the "done" column to your watched columns, or just select the "Show PRs in all columns" option, to see the changes in the done column. 

Does that solve the issue you were seeing? Let me know if not, thanks again.

**Review by sondoselayyanOP — APPROVED** [review]
LGTM!

**Review by jana-998 — APPROVED** [review]

**Comment by TomMokrzecki** [comment]
Works! Love my labels, ty!
QA approved, marked ready for FR

**Comment by hm-ysoliz** [comment]
URLs tested:
https://cds-global.atlassian.net/jira/software/c/projects/COLI/boards/391
https://cds-global.atlassian.net/jira/software/c/projects/COLI/boards/256
https://cds-global.atlassian.net/jira/software/c/projects/COTRO/boards/257
https://cds-global.atlassian.net/jira/software/c/projects/MOAPT/boards/275

Steps:

- Downloaded and unzipped file.
- Turned off existing Chrome extension (version 0.14)
- Turned on Enhanced version of JIRA
- Loaded unpacked extension into Chrome (version 0.14.1)
- Generated a new token using the instructions here (https://www.notion.so/mediaos/Github-Labels-in-Jira-b05f734721b8463f9b79adab7a2ba462) including SSO auth
- Updated extension within Options with new token and tested various iterations (hiding merged PRs, hiding closed PRs, Code Reviewers, and showing PRs in all columns)

Feature Review Passed ✅ 

image

**Review by hm-ysoliz — APPROVED** [review]
Feature Review Passed (in comments)

**Pull request merged**